ddl/partition.go

```diff
@@ -689,6 +701,182 @@ func onTruncateTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (int64, e
 	return ver, nil
 }
 
+// onExchangeTablePartition exchange partition data
+func (w *worker) onExchangeTablePartition(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
+	var (
+		ptSchemaID     int64
+		ptID           int64
+		partName       string
+		withValidation bool
+	)
+
+	if err := job.DecodeArgs(&ptSchemaID, &ptID, &partName, &withValidation); err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	ntDbInfo, err := t.GetDatabase(job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	nt, err := getTableInfoAndCancelFaultJob(t, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	pt, err := getTableInfo(t, ptID, ptSchemaID)
+	if err != nil {
+		if infoschema.ErrDatabaseNotExists.Equal(err) || infoschema.ErrTableNotExists.Equal(err) {
+			job.State = model.JobStateCancelled
+		}
+		return ver, errors.Trace(err)
+	}
+
+	err = checkExchangePartition(pt, nt)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	err = checkTableDefCompatible(pt, nt)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	if withValidation {
+		err = checkExchangePartitionRecordValidation(w, pt.Partition, partName, ntDbInfo.Name, nt.Name)
+		if err != nil {
+			job.State = model.JobStateCancelled
+			return ver, errors.Trace(err)
+		}
+	}
+
+	// partition table base auto id
+	ptBaseID, err := t.GetAutoTableID(ptSchemaID, pt.ID)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	// non-partition table base auto id
+	ntBaseID, err := t.GetAutoTableID(job.SchemaID, nt.ID)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	partDef, err := getPartitionDef(pt, partName)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	tempID := partDef.ID
+	// exchange table meta id
+	partDef.ID = nt.ID
+
+	err = t.UpdateTable(ptSchemaID, pt)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	// recreate non-partition table meta info
+	err = t.DropTableOrView(job.SchemaID, nt.ID, true)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	nt.ID = tempID
+
+	err = t.CreateTableOrView(job.SchemaID, nt)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+
+	_, err = t.GenAutoTableID(ptSchemaID, pt.ID, ntBaseID)
```

Comment:
If do not exchange auto_id, it would cause very bad effect on inserting data after table exhanged partition when both `pt` and `nt` have data. Due to `TIDB` have no part_auto_id like mysql(`I am not sure , maybe i am not found it`), I have no idea deal with this except echange the auto id of `pt` and `nt`. While the auto_id of  partition table is unregular. Sometimes it could generate `auto_id=30000` in partition table due to multi region(`I am not sure`)。If do not exchange auto_id. It could cause result as followers:
```sql
mysql> create table pt (a int) partition by range(a) (partition p0 values less than (3));
Query OK, 0 rows affected (0.07 sec)

mysql> create table nt(a int);
Query OK, 0 rows affected (0.06 sec)
mysql> insert into nt values (0), (1),(2);
Query OK, 3 rows affected (0.00 sec)
Records: 3  Duplicates: 0  Warnings: 0
mysql> insert into pt values (1);
Query OK, 1 row affected (0.00 sec)
mysql> alter table pt exchange partition p0 with table nt;
Query OK, 0 rows affected (0.06 sec)

mysql> select * from nt;
+------+
| a    |
+------+
|    1 |
+------+
1 row in set (0.00 sec)

mysql> insert into nt values (2);
Query OK, 1 row affected (0.01 sec)
mysql> select * from nt;
+------+
| a    |
+------+
|    2 |
+------+
1 row in set (0.00 sec)


```

Reply:
@zhaox1n If we don't exchange the `auto_id`, the allocated `tidb_rowid` could override the data in the other table. 

However, @tiancaiamao 's doubt is reasonable. The following case illustrates that if we simply exchange the `auto_id`, it could also lead to data corruption:
```sql
drop table if exists pt, nt;
create table pt (a int) auto_id_cache 1 partition by hash (a) partitions 2;
create table nt (a int) auto_id_cache 1;
insert into pt values (1), (2), (3);  -- set pt.next_global_rowid = 7
insert into nt values (1);  -- set nt.next_global_rowid = 3
alter table pt exchange partition p1 with table nt;
-- [table data] pt: 1 2 3
-- [table data] nt: 1
insert into pt values (5);
select * from pt; -- should contains 4 rows
-- [expected table data] pt: 1 2 3 5
-- [actual table data] pt: 1 2 5 (3 is override by 5)
```

Reply:
You can check the value of `next_global_rowid` by `show table nt next_row_id;`.

Reply:
After some expirements with MySQL, I find those two conclusion:

1. The auto increment id of the partitioned table should never decrease
2. The auto increment id of the normal table could be the largest record of the partition been exchanged

It means when a partitioned table `pt` exchanges a partition with a normal table `nt`, if the auto increment id of `nt` is greater than the `pt`, then the `pt` should update it's auto id, otherwise, `pt` keeps its auto id the original one.

Reply:
I think the conservative approach is to choose the max `auto_id`, and set it to both tables.

Reply:
So do i need to change the implementation of auto_id when exchange partition? or set the max one ? @tangenta

Reply:
How about this:
```go
ptStep := mathutil.MaxInt64(0, ntBaseID - ptBaseID)
_, err = t.GenAutoTableID(ptSchemaID, pt.ID, ptStep)
```
By the way, `auto_random` ID(if exists) should be done something similar.

Reply:
It's safe to add a if condition:

```
if ntBaseID > ptBaseID {
    // update ptBaseID at least greater than ntBaseID
} 
// else nothing need to be done
```

Reply:
It seems that row_id dose not effect on partition table.It would not show data curruption in partition table.   It's a feature?
```sql
drop table if exists pt, nt;
create table pt (a int) auto_id_cache 1 partition by hash (a) partitions 2;
create table nt (a int) auto_id_cache 1;
insert into pt values (2), (4), (6);  -- set pt.next_global_rowid = 7
insert into nt values (1),(3),(5), (7);  -- set nt.next_global_rowid = 9
alter table pt exchange partition p1 with table nt;
-- [table data] pt: [1,3,5,7,2,4,6] row_id = 9
-- [table data] nt: [] row_id = 9
insert into pt values (10);
select * from pt; 
-- [actual table data] pt: [1,3,5,7,2,4,6,10]  
```

Reply:
The comment above is updated and explains the reason.